pkg/orchestrator/scheduler/types.go

```diff
@@ -104,6 +106,109 @@ func (set execSet) filterFailed() execSet {
 	return set.filterByState(model.ExecutionStateFailed)
 }
 
+// filterOverSubscriptions partitions executions based on if they are more than the desired count.
+func (set execSet) filterByOverSubscriptions(desiredCount int) (execSet, execSet) {
+	remaining := make(execSet)
+	overSubscriptions := make(execSet)
+
+	count := 0
+	for _, exec := range set.ordered() {
+		if count >= desiredCount {
+			overSubscriptions[exec.ComputeReference] = exec
+		} else {
+			remaining[exec.ComputeReference] = exec
+		}
+		count++
+	}
+	return remaining, overSubscriptions
+}
+
+// filterByNodeHealth partitions executions based on their node's health status.
+func (set execSet) filterByNodeHealth(nodeInfos map[string]*model.NodeInfo) (execSet, execSet) {
+	healthy := make(execSet)
+	lost := make(execSet)
+
+	for _, exec := range set {
+		if _, ok := nodeInfos[exec.NodeID]; !ok {
+			lost[exec.ComputeReference] = exec
+			log.Debug().Msgf("Execution %s is running on node %s which is not healthy", exec.ComputeReference, exec.NodeID)
+		} else {
+			healthy[exec.ComputeReference] = exec
+		}
+	}
+	return healthy, lost
+}
+
+// executionsByApprovalStatus represents the different sets of executions based on their approval status.
+type executionsByApprovalStatus struct {
+	running   execSet
+	toApprove execSet
+	toReject  execSet
+	pending   execSet
+}
+
+// activeCount returns the number of active executions, excluding rejected ones.
+func (e executionsByApprovalStatus) activeCount() int {
+	return len(e.running) + len(e.toApprove) + len(e.pending)
+}
+
+// filterByApprovalStatus partitions executions based on their approval status.
+func (set execSet) filterByApprovalStatus(desiredCount int) executionsByApprovalStatus {
+	nonTermExecs := set.filterNonTerminal()
+	running := nonTermExecs.filterRunning()
+	toApprove := make(execSet)
+	toReject := make(execSet)
+	pending := make(execSet)
+
+	//TODO: we are approving the oldest executions first, we should probably
+	// approve the ones with highest rank first
```

Comment:
What is the rank in this context?

Reply:
Nodes are ranked when creating executions. We should persist the rank that was used when selecting nodes in the execution itself, and then discard executions with lower rank if we need to discard some